sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -54,7 +54,7 @@ object CollationTypeCasts extends TypeCoercionRule {
 
     case otherExpr @ (
       _: In | _: InSubquery | _: CreateArray | _: ArrayJoin | _: Concat | _: Greatest | _: Least |
-      _: Coalesce | _: BinaryExpression | _: ConcatWs) =>
+      _: Coalesce | _: BinaryExpression | _: ConcatWs | _: StringLPad | _: StringRPad) =>
```

Comment:
@uros-db  That is not so, in my understanding.

`collateToSingleType` calls  `getOutputCollation`  and then does     `exprs.map(e => castStringType(e, st).getOrElse(e))`

`getOutputCollation`, according to the function comment blocks (verified by me looking at the code), "will only be affected by collated StringTypes or complex DataTypes with collated StringTypes (e.g. ArrayType)"


`castStringType` , according to the function comment blocks (verified by me looking at the code), "casts given expression to collated StringType with id equal to collationId only if expression has StringType in the first place."

Reply:
unfortunately, CollationTypeCasts doesn't quite do what you would expect based on those descriptions - we're working on fixing that

but if you were to run this:
```
  test("Support StringLPad string expressions with collation") {
    val query = "SELECT lpad('abc', collate('5', 'unicode_ci'), ' ')"
    checkAnswer(sql(query), Row("  abc"))
    assert(sql(query).schema.fields.head.dataType.sameType(StringType(3)))
  }
```
  you would find that resulting datatype is UNICODE_CI (instead of UTF8_BINARY which should be the correct type)
  
  the reason for this behaviour is found in the ordering of rules in `TypeCoercion` - here you'll find that for example `CollationTypeCasts` comes before `FunctionArgumentConversion`, so in the above example collation is taken from '5' before it's converted into a 5

Reply:
I see.

Reply:
@uros-db done. please re-review